src/server/worker/pipeline/transform/worker_v2.go

```diff
@@ -0,0 +1,93 @@
+package transform
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"github.com/pachyderm/pachyderm/src/server/worker/datum"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+)
+
+// WorkerV2 handles a transform pipeline work subtask, then returns.
+// TODO:
+// datum queuing (probably should be handled by datum package).
+// s3 input / gateway stuff (need more information here).
+// spouts.
+// joins.
+// capture datum logs.
+// file download features (empty / lazy files). Need to check over the pipe logic.
+// git inputs.
+// handle custom user set for execution.
+// Taking advantage of symlinks during upload?
+func WorkerV2(driver driver.Driver, logger logs.TaggedLogger, subtask *work.Task, status *Status) (retErr error) {
+	datumSet, err := deserializeDatumSetV2(subtask.Data)
+	if err != nil {
+		return err
+	}
+	return status.withJob(datumSet.JobID, func() error {
+		logger = logger.WithJob(datumSet.JobID)
+		if err := logger.LogStep("datum task", func() error {
+			return handleDatumSetV2(driver, logger, datumSet, status)
+		}); err != nil {
+			return err
+		}
+		subtask.Data, err = serializeDatumSetV2(datumSet)
+		return err
+	})
+}
+
+// TODO: It would probably be better to write the output to temporary file sets and expose an operation through pfs for adding a temporary fileset to a commit.
+func handleDatumSetV2(driver driver.Driver, logger logs.TaggedLogger, datumSet *DatumSetV2, status *Status) error {
+	pachClient := driver.PachClient()
+	storageRoot := filepath.Join(driver.InputDir(), client.PPSScratchSpace, uuid.NewWithoutDashes())
+	datumSet.Stats = &datum.Stats{ProcessStats: &pps.ProcessStats{}}
+	// Setup file operation client for output meta commit.
+	metaCommit := datumSet.MetaCommit
+	return pachClient.WithFileOperationClientV2(metaCommit.Repo.Name, metaCommit.ID, func(focMeta *client.FileOperationClient) error {
+		// Setup file operation client for output PFS commit.
+		outputCommit := datumSet.OutputCommit
+		return pachClient.WithFileOperationClientV2(outputCommit.Repo.Name, outputCommit.ID, func(focPFS *client.FileOperationClient) error {
+			// Setup datum set for processing.
+			return datum.WithSet(pachClient, storageRoot, func(s *datum.Set) error {
+				di := datum.NewFileSetIterator(pachClient, client.TmpRepoName, datumSet.FileSet)
+				// Process each datum in the assigned datum set.
+				return di.Iterate(func(meta *datum.Meta) error {
```

Comment:
I don't think it would be too hard to parallelize as is because we would just have a goroutine get kicked off by `WithDatum` and managed by the set (we would probably need a new context for the set though to know when we should exit the iteration, on that note I should probably add a context to the iterator interface), but what I want to do long term is in line with what you suggested which is capture more of this in the datum package.